src/test/java/net/snowflake/client/jdbc/ResultSetStructuredTypesLatestIT.java

```diff
@@ -0,0 +1,167 @@
+import java.sql.Connection;
+import java.sql.Date;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import net.snowflake.client.ConditionalIgnoreRule;
+import net.snowflake.client.RunningOnGithubAction;
+import net.snowflake.client.category.TestCategoryStructuredType;
+import net.snowflake.client.core.structs.SnowflakeObjectTypeFactories;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+@Category(TestCategoryStructuredType.class)
+public class ResultSetStructuredTypesLatestIT extends BaseJDBCTest {
+  private final String queryResultFormat;
+
+  public ResultSetStructuredTypesLatestIT() {
+    this("JSON");
+  }
+
+  protected ResultSetStructuredTypesLatestIT(String queryResultFormat) {
+    this.queryResultFormat = queryResultFormat;
+  }
+
+  public Connection init() throws SQLException {
+    Connection conn = BaseJDBCTest.getConnection(BaseJDBCTest.DONT_INJECT_SOCKET_TIMEOUT);
+    try (Statement stmt = conn.createStatement()) {
+      stmt.execute("alter session set ENABLE_STRUCTURED_TYPES_IN_CLIENT_RESPONSE = true");
+      stmt.execute("alter session set IGNORE_CLIENT_VESRION_IN_STRUCTURED_TYPES_RESPONSE = true");
+      stmt.execute("alter session set jdbc_query_result_format = '" + queryResultFormat + "'");
+    }
+    return conn;
+  }
+
+  // TODO Structured types feature exists only on QA environments
+  @Test
+  @ConditionalIgnoreRule.ConditionalIgnore(condition = RunningOnGithubAction.class)
+  public void testMapStructToObjectWithFactory() throws SQLException {
+    testMapJson(true);
+  }
+
+  // TODO Structured types feature exists only on QA environments
+  @Test
+  @ConditionalIgnoreRule.ConditionalIgnore(condition = RunningOnGithubAction.class)
+  public void testMapStructToObjectWithReflection() throws SQLException {
+    testMapJson(false);
+    testMapJson(true);
+  }
+
+  private void testMapJson(boolean registerFactory) throws SQLException {
+    if (registerFactory) {
+      SnowflakeObjectTypeFactories.register(SimpleClass.class, SimpleClass::new);
+    } else {
+      SnowflakeObjectTypeFactories.unregister(SimpleClass.class);
+    }
+    try (Connection connection = init();
+        Statement statement = connection.createStatement();
+        ResultSet resultSet =
+            statement.executeQuery("select {'string':'a'}::OBJECT(string VARCHAR)"); ) {
+      resultSet.next();
+      SimpleClass object = resultSet.getObject(1, SimpleClass.class);
+      assertEquals("a", object.getString());
+    }
+  }
+
+  // TODO Structured types feature exists only on QA environments
+  @Test
+  @ConditionalIgnoreRule.ConditionalIgnore(condition = RunningOnGithubAction.class)
+  public void testMapAllTypesOfFields() throws SQLException {
+    SnowflakeObjectTypeFactories.register(AllTypesClass.class, AllTypesClass::new);
```

Comment:
why don't we jave logic for registering or not classes also there?

Reply:
Added

Reply:
we are register it and in the next lines we have conditional logic for the same - it should be made once in this method

Reply:
Fixed